File: Games/Easy/Connect4/connect4.py
import numpy as np
import pygame
import sys
import math
from typing import Tuple, List, Optional

# --- Constants ---
# Colors
BLUE = (20, 50, 200)
BLACK = (0, 0, 0)
RED = (210, 40, 40)
YELLOW = (255, 200, 0)

# Board dimensions
ROW_COUNT = 6
COLUMN_COUNT = 7
WINDOW_LENGTH = 4

# Player pieces
PLAYER_1_PIECE = 1
PLAYER_2_PIECE = 2

class Board:
    """Manages the Connect 4 board state and logic."""
    def __init__(self):
        self.board = np.zeros((ROW_COUNT, COLUMN_COUNT), dtype=int)

    def drop_piece(self, row: int, col: int, piece: int):
        self.board[row][col] = piece

    def is_valid_location(self, col: int) -> bool:
        return self.board[ROW_COUNT - 1][col] == 0

    def get_next_open_row(self, col: int) -> Optional[int]:
        for r in range(ROW_COUNT):
            if self.board[r][col] == 0:
                return r
        return None

    def winning_move(self, piece: int) -> bool:
        # Check all possible winning combinations more concisely
        # Horizontal
        for c in range(COLUMN_COUNT - 3):
            for r in range(ROW_COUNT):
                if all(self.board[r][c + i] == piece for i in range(WINDOW_LENGTH)):
                    return True
        # Vertical
        for c in range(COLUMN_COUNT):
            for r in range(ROW_COUNT - 3):
                if all(self.board[r + i][c] == piece for i in range(WINDOW_LENGTH)):
                    return True
        # Positively sloped diagonals
        for c in range(COLUMN_COUNT - 3):
            for r in range(ROW_COUNT - 3):
                if all(self.board[r + i][c + i] == piece for i in range(WINDOW_LENGTH)):
                    return True
        # Negatively sloped diagonals
        for c in range(COLUMN_COUNT - 3):
            for r in range(3, ROW_COUNT):
                if all(self.board[r - i][c + i] == piece for i in range(WINDOW_LENGTH)):
                    return True
        return False

    def is_full(self) -> bool:
        return 0 not in self.board[ROW_COUNT - 1]

class Connect4Game:
    """Manages the main game loop, graphics, and player interactions for a two-player game."""
    def __init__(self, square_size: int = 100):
        pygame.init()
        self.square_size = square_size
        self.width = COLUMN_COUNT * self.square_size
        self.height = (ROW_COUNT + 1) * self.square_size
        self.radius = int(self.square_size / 2 - 5)

        self.screen = pygame.display.set_mode((self.width, self.height))
        pygame.display.set_caption("Connect 4 - Two Player")
        self.font = pygame.font.SysFont("monospace", 75)

        self.board = Board()
        self.game_over = False
        self.turn = 0  # 0 for Player 1, 1 for Player 2

    def draw_board(self):
        self.screen.fill(BLACK)
        for c in range(COLUMN_COUNT):
            for r in range(ROW_COUNT):
                pygame.draw.rect(self.screen, BLUE, (c * self.square_size, r * self.square_size + self.square_size, self.square_size, self.square_size))
                pygame.draw.circle(self.screen, BLACK, (int(c * self.square_size + self.square_size / 2), int(r * self.square_size + self.square_size + self.square_size / 2)), self.radius)

        for c in range(COLUMN_COUNT):
            for r in range(ROW_COUNT):
                if self.board.board[r][c] == PLAYER_1_PIECE:
                    pygame.draw.circle(self.screen, RED, (int(c * self.square_size + self.square_size / 2), self.height - int(r * self.square_size + self.square_size / 2)), self.radius)
                elif self.board.board[r][c] == PLAYER_2_PIECE:
                    pygame.draw.circle(self.screen, YELLOW, (int(c * self.square_size + self.square_size / 2), self.height - int(r * self.square_size + self.square_size / 2)), self.radius)

        # Draw the current player's piece at the top
        if not self.game_over:
            color = RED if self.turn == 0 else YELLOW
            posx = pygame.mouse.get_pos()[0]
            pygame.draw.circle(self.screen, color, (posx, int(self.square_size / 2)), self.radius)

        pygame.display.update()

    def handle_mouse_click(self, posx: int):
        if self.game_over:
            return

        col = posx // self.square_size
        if self.board.is_valid_location(col):
            row = self.board.get_next_open_row(col)
            piece = PLAYER_1_PIECE if self.turn == 0 else PLAYER_2_PIECE
            self.board.drop_piece(row, col, piece)

            if self.board.winning_move(piece):
                player_num = 1 if self.turn == 0 else 2
                color = RED if player_num == 1 else YELLOW
                self.end_game(f"Player {player_num} wins!!", color)
            elif self.board.is_full():
                self.end_game("It's a Tie!", BLUE)

            self.turn = (self.turn + 1) % 2

    def end_game(self, message: str, color: Tuple[int, int, int]):
        label = self.font.render(message, True, color)
        self.screen.blit(label, (40, 10))
        self.game_over = True

    def run(self):
        while True:
            for event in pygame.event.get():
                if event.type == pygame.QUIT:
                    sys.exit()
                if event.type == pygame.MOUSEBUTTONDOWN:
                    self.handle_mouse_click(event.pos[0])

            self.draw_board()

            if self.game_over:
                pygame.time.wait(3000)
                # Optional: break the loop to close the window after delay
                # break
                # Or reset the game:
                # self.board = Board()
                # self.game_over = False
                # self.turn = 0
                sys.exit() # For simplicity, exit after game over

if __name__ == "__main__":
    game = Connect4Game()
    game.run()

File: Games/Easy/Connect4/connect4.java
package Games.Easy.Connect4;

import java.util.Scanner;

public class connect4 {
  public static void main(String[] args) {
    try (Scanner inp = new Scanner(System.in)) {
      char[][] board = new char[6][7];

      for (int i = 0; i < 6; i++) {
        for (int j = 0; j < 7; j++) {
          board[i][j] = '.';
        }
      }

      int move = 1;
      char player = 'A';
      boolean gameOver = false;

      while (!gameOver && move <= 42) {
        boolean validMove = false;
        int play;
        do {
          display(board);
          System.out.println("Player " + player +
              ": Enter a column number (1-7): ");
          play = inp.nextInt();

          if (play < 1 || play > 7) {
            System.out.println("Invalid column number. Try again.");
          } else {
            validMove = true;
          }
        } while (!validMove);

        for (int row = 5; row >= 0; row--) {
          if (board[row][play - 1] == '.') {
            board[row][play - 1] = player;
            break;
          }
        }
        gameOver = checkWinner(player, board);

        if (player == 'A') {
          player = 'B';
        } else {
          player = 'A';
        }
        move++;
      }
      display(board);

      if (gameOver) {
        if (player == 'A') {
          System.out.println("Player B wins!");
        } else {
          System.out.println("Player A wins!");
        }
      } else {
        System.out.println("It's a tie!");
      }
    }
  }

  private static boolean checkWinner(char player, char[][] board) {
    boolean winner = false;

    for (int row = 0; row < 6; row++) {
      for (int col = 0; col < 7; col++) {
        if (board[row][col] == player) {
          if (row + 3 < 6) {
            if (board[row][col] == board[row + 1][col] &&
                board[row][col] == board[row + 2][col] &&
                board[row][col] == board[row + 3][col]) {
              winner = true;
            }
          }
          if (col + 3 < 7) {
            if (board[row][col] == board[row][col + 1] &&
                board[row][col] == board[row][col + 2] &&
                board[row][col] == board[row][col + 3]) {
              winner = true;
            }
          }
          if (row + 3 < 6 && col + 3 < 7) {
            if (board[row][col] == board[row + 1][col + 1] &&
                board[row][col] == board[row + 2][col + 2] &&
                board[row][col] == board[row + 3][col + 3]) {
              winner = true;
            }
          }
          if (row - 3 >= 0 && col + 3 < 7) {
            if (board[row][col] == board[row - 1][col + 1] &&
                board[row][col] == board[row - 2][col + 2] &&
                board[row][col] == board[row - 3][col + 3]) {
              winner = true;
            }
          }
        }
      }
    }
    return winner;
  }

  public static void display(char[][] board) {
    System.out.println(" 0 1 2 3 4 5 6");
    System.out.println("---------------");
    for (int row = 0; row < 6; row++) {
      System.out.print(row + "|");
      for (int col = 0; col < 7; col++) {
        System.out.print(board[row][col] + " ");
      }
      System.out.println();
      System.out.println("---------------");
    }
  }
}